### DISCUSSION TITLE
'close' function is called although data is remained in sendbuffer

### DISCUSSION THREAD

**Discussion by rubidam** [ANLAB-KAIST/KENSv3#122] [Project Q&A 2. Transport Layer (TCP)] [answered]
I have a question about how the close and write functions work.

Among the 'TestTransfer' test cases, 4 cases related to 'send_symmetric', 'send_EOF' are failing. In all four cases, the difference between 'expect_size' and 'total_size' is exactly 1612, and I found that 'close' function is called although this 1612 bytes of data is remained in 'sendbuffer.

**My question is, is it not wrong to call 'close' although data is in 'sendbuffer', and if data remains in 'rcvbuffer' or 'sendbuffer' when the 'close' function is called, do I have to handle it before sending fin packet?
Or did I write the 'write' function and 'packetarrived' function wrongly, so the above situation is created?**

-----------------------------------------------------------------------------------------------------------------------------------------------
close 함수와 write 함수의 작동방식에 대해 질문이 있습니다.

'TestTransfer' 테스트 케이스 중에서 'send_symmetric', 'send_EOF' 과 관련된 4개의 케이스가 실패하고 있습니다. 4개의 케이스 모두 'expect_size'와 'total_size'가 정확히 1612가 차이가 나며, 이 1612byte의 데이터가 'sendbuffer'에 담겨있는 채로 'close'가 불려 코드가 끝난다는 것을 알게 되었습니다. 

**제 질문은 'sendbuffer'에 데이터가 담겨있는 채로 'close'가 불리는 상황이 틀린 상황이 아니며 'close' 함수가 불릴때 'rcvbuffer'나 'sendbuffer'에 데이터가 남아있으면 fin 패킷을 보내기전에 데이터를 처리해줘야 하는겁니까?
아니면 제가 작성한 'write' 함수와 'packetarrived' 함수를 잘못 작성하여 위와 같은 상황이 만들어진겁니까?**

**Comment by whkim0** [comment, marked as answer]
> My question is, is it not wrong to call 'close' although data is in 'sendbuffer'

No, it is not wrong. Calling `close()` means "I have no more data to send.", rather than "I have no more data in 'sendbuffer'."
(But depends on your implementation)

> If data remains in 'rcvbuffer' or 'sendbuffer' when the 'close' function is called, do I have to handle it before sending fin packet?

Yes, TCP should reliably send any data that is already queued to be sent to the other end.

> Or did I write the 'write' function and 'packetarrived' function wrongly, so the above situation is created?

I'm not sure how you implemented handling a packet and closing a connection, but your TCP implementation should try to send any data in a buffer.

I recommend you to read this document: [RFC793, page 37](https://www.rfc-editor.org/rfc/rfc793#page-37).

**Reply by rubidam** [reply]
Thank you for your answer!
